File: 2024/python/src/day1.py
from solution import Solution
from typing import List, Tuple


class Day1(Solution):
    def first_part(self, input_file: str):
        file = open(input_file, "r")
        input = file.read()
        left, right = parse_input_lists(input)
        output = sum(
            map(lambda tup: abs(tup[0] - tup[1]), zip(sorted(left), sorted(right)))
        )
        print("Difference", output)

    def second_part(self, input_file: str):
        file = open(input_file, "r")
        input = file.read()
        left, right = parse_input_lists(input)

        similarity_score = sum(map(lambda x: x * len(list(filter(lambda y: y == x, right))), left))
        print("Similarity score: ", similarity_score)


def parse_input_lists(input: str) -> Tuple[List[int], List[int]]:
    left = []
    right = []
    for line in input.splitlines():
        first, second = line.split("   ")
        left.append(int(first))
        right.append(int(second))

    return (left, right)

File: 2024/java/src/main/java/solutions/day1/Day1.java
package solutions.day1;

import com.google.common.collect.Streams;
import org.apache.commons.lang3.tuple.Pair;
import solutions.Solution;
import solutions.Utils;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Stream;

public class Day1 implements Solution {
    @Override
    public void firstPart(String inputFile) {
        Pair<List<Integer>, List<Integer>> inputLists = parseInputLists(
                Utils.readInputAsStream(inputFile));
        var sortedInputLists = Pair.of(inputLists.getLeft().stream().sorted().toList(),
                inputLists.getRight().stream().sorted().toList());

        int difference = Streams.zip(sortedInputLists.getLeft().stream(),
                        sortedInputLists.getRight().stream(), Pair::of)
                .map(pair -> Math.abs(pair.getLeft() - pair.getRight()))
                .reduce(Integer::sum)
                .get();

        System.out.printf("Difference: %s\n", difference);
    }

    @Override
    public void secondPart(String inputFile) {
        Pair<List<Integer>, List<Integer>> inputLists = parseInputLists(
                Utils.readInputAsStream(inputFile));
        Map<Integer, Integer> counts = new HashMap<>();
        for (int x : inputLists.getLeft()) {
            counts.put(x, (int) inputLists.getRight().stream().filter(y -> y == x).count());
        }
        int similarityScore = counts.entrySet()
                .stream()
                .map(entry -> entry.getKey() * entry.getValue())
                .reduce(Integer::sum)
                .get();

        System.out.printf("Similarity score: %s\n", similarityScore);
    }


    private Pair<List<Integer>, List<Integer>> parseInputLists(Stream<String> input) {
        List<Integer> left = new ArrayList<>();
        List<Integer> right = new ArrayList<>();

        input.forEach(line -> {
            var parts = line.split("   ");
            assert (parts.length == 2) : "Each line in the input should contain entries " + "for "
                    + "both lists.";
            left.add(Integer.parseInt(parts[0]));
            right.add(Integer.parseInt(parts[1]));
        });
        return Pair.of(left, right);
    }

}
